include/oboe/AudioStreamCallback.h

```diff
@@ -75,49 +76,106 @@ class AudioStreamCallback {
      * If you need to move data, eg. MIDI commands, in or out of the callback function then
      * we recommend the use of non-blocking techniques such as an atomic FIFO.
      *
-     * @param oboeStream pointer to the associated stream
+     * @param audioStream pointer to the associated stream
      * @param audioData buffer containing input data or a place to put output data
      * @param numFrames number of frames to be processed
      * @return DataCallbackResult::Continue or DataCallbackResult::Stop
      */
     virtual DataCallbackResult onAudioReady(
-            AudioStream *oboeStream,
+            AudioStream *audioStream,
             void *audioData,
             int32_t numFrames) = 0;
+};
+
+/**
+ * AudioStreamDataCallback defines a callback interface for
+ * being alerted when a stream has an error or is disconnected
+ * using `onError*` methods.
+ *
+ * It is used with AudioStreamBuilder::setErrorCallback().
+ */
+class AudioStreamErrorCallback {
+public:
+    virtual ~AudioStreamErrorCallback() = default;
 
     /**
-     * This will be called when an error occurs on a stream or when the stream is disconnected.
+     * This will be called first when an error occurs on a stream or when the stream is disconnected.
+     *
+     * It can be used to override and customize the normal error processing.
+     * Use of this method is considered an advanced technique.
+     * It might, for example, be used if an app want to use a high level lock when
+     * closing and reopening a stream.
+     * Or it might be used when an app want to signal a management thread that handles
+     * all of the stream state.
+     *
+     * If this method returns false then
+     * the stream will be stopped, and onErrorBeforeClose() will be called,
+     * then the stream will be closed and onErrorAfterClose() will be closed.
      *
-     * Note that this will be called on a different thread than the onAudioReady() thread.
-     * This thread will be created by Oboe.
+     * If this method returns true then the normal error processing will not occur.
+     * In that case, the app MUST stop() and close() the stream!
+     *
+     * Note that this will be called on a thread created by Oboe.
+     *
+     * @param audioStream pointer to the associated stream
+     * @param error
+     */
+    virtual bool onError(AudioStream* /* audioStream */, Result /* error */) {
+        return false; // false means the stream will be stopped and closed by Oboe
+        // return true; // true means the stream will be stopped and closed by the app
```

Comment:
There's some redundancy in the comments here. Consider consolidating this into the `@return` comment or removing this comment and the one above altogether.

Reply:
OK, will do in a subsequent patch.